File: web/src/api/es-index-pattern.ts
import request from './request'
import type { ApiResponse } from '@/types'

export interface ESIndexPattern {
  id: number
  datasource_id: number
  name: string
  time_field: string
  allow_hide_system_indices: boolean
  fields_format: string
  cross_cluster_enabled: boolean
  note: string
  created_by: string
  updated_by: string
  created_at: string
  updated_at: string
}

export interface CreateESIndexPatternRequest {
  datasource_id: number
  name: string
  time_field?: string
  allow_hide_system_indices?: boolean
  fields_format?: string
  cross_cluster_enabled?: boolean
  note?: string
}

export type UpdateESIndexPatternRequest = CreateESIndexPatternRequest

export const esIndexPatternApi = {
  list: (params?: { datasource_id?: number }) =>
    request.get<ApiResponse<ESIndexPattern[]>>('/es-index-patterns', { params }),

  get: (id: number) =>
    request.get<ApiResponse<ESIndexPattern>>(`/es-index-patterns/${id}`),

  create: (data: CreateESIndexPatternRequest) =>
    request.post<ApiResponse<ESIndexPattern>>('/es-index-patterns', data),

  update: (id: number, data: UpdateESIndexPatternRequest) =>
    request.put<ApiResponse<null>>(`/es-index-patterns/${id}`, data),

  delete: (id: number) =>
    request.delete<ApiResponse<null>>(`/es-index-patterns/${id}`),
}

File: internal/model/es_index_pattern.go
package model

import (
	"fmt"

	"gorm.io/gorm"
)

// ESIndexPattern stores an Elasticsearch index pattern configuration
// associated with a datasource.
type ESIndexPattern struct {
	gorm.Model

	DatasourceID           uint   `json:"datasource_id" gorm:"not null;index:idx_ds_name,unique"`
	Name                   string `json:"name" gorm:"size:191;not null;index:idx_ds_name,unique"`
	TimeField              string `json:"time_field" gorm:"size:128;not null;default:'@timestamp'"`
	AllowHideSystemIndices bool   `json:"allow_hide_system_indices" gorm:"default:false"`
	FieldsFormat           string `json:"fields_format" gorm:"type:text"` // JSON
	CrossClusterEnabled    bool   `json:"cross_cluster_enabled" gorm:"default:false"`
	Note                   string `json:"note" gorm:"size:512"`
	CreatedBy              string `json:"created_by" gorm:"size:64"`
	UpdatedBy              string `json:"updated_by" gorm:"size:64"`
}

func (ESIndexPattern) TableName() string { return "es_index_patterns" }

// Verify validates required fields.
func (p *ESIndexPattern) Verify() error {
	if p.Name == "" {
		return fmt.Errorf("name is required")
	}
	if p.DatasourceID == 0 {
		return fmt.Errorf("datasource_id is required")
	}
	return nil
}
